azuredevops.go

```diff
@@ -276,202 +160,388 @@ func runAzureDevOpsHandler(c *gin.Context) {
+			provider, err = client.PostProvider(
+				wharfapi.Provider{Name: apiProviderName, URL: i.URL, TokenID: token.TokenID})
+			if err != nil {
+				fmt.Println("Unable to post provider: ", err)
+				ginutil.WriteAPIClientWriteError(c, err,
+					fmt.Sprintf("Unable to get or create provider from '%s'.", i.URL))
+				return wharfapi.Provider{}, err
+			}
+		}
+	}
+
+	i.URL = provider.URL
+	return provider, nil
+}
+
+func getRepositoriesWritesError(c *gin.Context, i importBody, project azureDevOpsProject) (azureDevOpsRepositoryResponse, error) {
+	const apiVersion string = "5.0"
+
+	urlPath, err := requests.ConstructGetURL(i.URL, map[string][]string{
+		"api-version": {apiVersion},
+	}, "%v/%v/%v", i.Group, project.Name, apiRepositories)
 	if err != nil {
 		fmt.Println("Unable to get url: ", err)
-		return []byte{}, err
+		ginutil.WriteInvalidParamError(c, err, "URL", fmt.Sprintf("Unable to parse URL %q", i.URL))
+		return azureDevOpsRepositoryResponse{}, err
 	}
 
-	urlPath.Path = fmt.Sprintf("%v/%v/%v", i.Group, project, apiRepositories)
-	data := url.Values{}
-	data.Set("api-version", "5.0")
-	urlPath.RawQuery = data.Encode()
 	fmt.Println(urlPath.String())
 
-	return getBodyFromRequest(i.User, i.Token, urlPath.String())
+	repositories := azureDevOpsRepositoryResponse{}
+	err = requests.GetAndParseJSON(&repositories, i.User, i.Token, urlPath)
+	if err != nil {
+		fmt.Println("Unable to get project repository: ", err)
+		ginutil.WriteProviderResponseError(c, err, "Could be caused by invalid JSON data structure." +
+			"\nMight be the result of an incompatible version of Azure DevOps.")
+		return azureDevOpsRepositoryResponse{}, err
+	}
+
+	if repositories.Count != 1 {
+		fmt.Println("One repository is required.")
+		ginutil.WriteAPIClientReadError(c, err,
+			fmt.Sprintf("There were %v repositories, we need it to be 1.",
+				repositories.Count))
+		return azureDevOpsRepositoryResponse{}, fmt.Errorf("one repository is required")
+	}
+
+	if repositories.Value[0].Project.ID != project.ID {
+		fmt.Println("Repository is not connected with project.")
+		ginutil.WriteAPIClientReadError(c, err,
+			fmt.Sprintf("Repository id (%v) and project id (%v) mismatch",
+				repositories.Value[0].Project.ID,
+				project.ID))
+		return azureDevOpsRepositoryResponse{}, fmt.Errorf("repository is not connected with project")
+	}
+
+	return repositories, nil
 }
 
-func getAzureDevOpsBuildDefinition(i importBody, project string) (string, error) {
-	urlPath, err := url.Parse(i.URL)
+func getBuildDefinitionWritesError(c *gin.Context, i importBody, projectName string) (string, error) {
+	urlPath, err := requests.ConstructGetURL(i.URL, map[string][]string{
+		"scopePath": {fmt.Sprintf("/%v", buildDefinitionFileName)},
+	}, "%s/%s/%s/%s/%s", i.Group, i.Project, apiRepositories, projectName, itemsPath)
 	if err != nil {
 		fmt.Println("Unable to get url: ", err)
+		ginutil.WriteInvalidParamError(c, err, "URL", fmt.Sprintf("Unable to parse URL %q", i.URL))
```

Comment:
nit: Use the JSON names for parameter problems

```suggestion
		ginutil.WriteInvalidParamError(c, err, "url", fmt.Sprintf("Unable to parse URL %q", i.URL))
```